File: langs/csharp/Adapter.cs
using System.Reflection;
using System.Text;
using System.Text.Json;

public class Adapter
{
    public static void Main(string[] args)
    {
        List<string> positional = args.Where(item => item.Length > 0 && item != "--" && !item.StartsWith('-')).ToList();
        if (positional.Count < 2)
        {
            Console.Error.WriteLine("usage: honepadrun cases.json ClassName");
            Environment.Exit(2);
        }

        string casesPath = positional[0];
        string className = positional[1];
        using JsonDocument doc = JsonDocument.Parse(File.ReadAllText(casesPath));
        if (doc.RootElement.ValueKind != JsonValueKind.Array)
        {
            throw new InvalidOperationException("cases.json must be a JSON list");
        }

        Type type =
            typeof(Adapter).Assembly.GetTypes().FirstOrDefault(item => item.Name == className)
            ?? throw new InvalidOperationException("missing type " + className);

        List<Dictionary<string, object?>> failed = new();
        int passed = 0;
        foreach (JsonElement row in doc.RootElement.EnumerateArray())
        {
            object obj =
                Activator.CreateInstance(type)
                ?? throw new InvalidOperationException("could not construct " + className);
            string caseId = row.GetProperty("id").GetString() ?? "";
            JsonElement calls = row.GetProperty("calls");
            bool ok = true;
            int index = 0;
            foreach (JsonElement call in calls.EnumerateArray())
            {
                string methodSnake = call.GetProperty("m").GetString() ?? "";
                string name = ToPascal(methodSnake);
                List<JsonElement> argv = call.GetProperty("a").EnumerateArray().ToList();
                JsonElement expected = call.GetProperty("e");
                object? actual;
                try
                {
                    actual = Invoke(obj, name, argv);
                }
                catch (Exception exc)
                {
                    failed.Add(FailRow(caseId, index, methodSnake, expected, "exc:" + Unwrap(exc).GetType().Name));
                    ok = false;
                    break;
                }
                if (!JsonEqual(actual, expected))
                {
                    failed.Add(FailRow(caseId, index, methodSnake, expected, actual));
                    ok = false;
                    break;
                }
                index += 1;
            }
            if (ok)
            {
                passed += 1;
            }
        }

        Dictionary<string, object?> report = new()
        {
            ["passed"] = passed,
            ["failed"] = failed,
        };
        Console.WriteLine(JsonSerializer.Serialize(report));
        Environment.Exit(failed.Count == 0 ? 0 : 1);
    }

    static Dictionary<string, object?> FailRow(
        string caseId,
        int index,
        string method,
        JsonElement expected,
        object? actual
    )
    {
        return new Dictionary<string, object?>
        {
            ["case"] = caseId,
            ["index"] = index,
            ["method"] = method,
            ["expected"] = expected,
            ["actual"] = actual,
        };
    }

    static string ToPascal(string snake)
    {
        string[] parts = snake.Split('_');
        StringBuilder output = new();
        foreach (string part in parts)
        {
            if (part.Length == 0)
            {
                continue;
            }
            output.Append(char.ToUpperInvariant(part[0]));
            if (part.Length > 1)
            {
                output.Append(part[1..]);
            }
        }
        return output.ToString();
    }

    static object? Invoke(object obj, string name, List<JsonElement> argv)
    {
        MethodInfo method =
            FindMethod(obj.GetType(), name, argv.Count)
            ?? throw new MissingMethodException(name);
        ParameterInfo[] types = method.GetParameters();
        object?[] converted = new object?[argv.Count];
        for (int i = 0; i < argv.Count; i++)
        {
            converted[i] = ConvertArg(argv[i], types[i].ParameterType);
        }
        return method.Invoke(obj, converted);
    }

    static MethodInfo? FindMethod(Type type, string name, int argc)
    {
        foreach (MethodInfo method in type.GetMethods())
        {
            if (method.Name == name && method.GetParameters().Length == argc)
            {
                return method;
            }
        }
        return null;
    }

    static object? ConvertArg(JsonElement arg, Type dest)
    {
        Type inner = Nullable.GetUnderlyingType(dest) ?? dest;
        if (arg.ValueKind == JsonValueKind.Null)
        {
            if (dest.IsValueType && Nullable.GetUnderlyingType(dest) == null)
            {
                throw new ArgumentException("null for non-nullable " + dest.Name);
            }
            return null;
        }
        if (inner == typeof(string))
        {
            if (arg.ValueKind != JsonValueKind.String)
            {
                throw new ArgumentException("cannot convert " + arg.ValueKind + " to string");
            }
            return arg.GetString();
        }
        if (inner == typeof(int))
        {
            if (arg.ValueKind != JsonValueKind.Number)
            {
                throw new ArgumentException("cannot convert " + arg.ValueKind + " to int");
            }
            return arg.GetInt32();
        }
        if (inner == typeof(long))
        {
            if (arg.ValueKind != JsonValueKind.Number)
            {
                throw new ArgumentException("cannot convert " + arg.ValueKind + " to long");
            }
            return arg.GetInt64();
        }
        if (inner == typeof(bool))
        {
            if (arg.ValueKind is not (JsonValueKind.True or JsonValueKind.False))
            {
                throw new ArgumentException("cannot convert " + arg.ValueKind + " to bool");
            }
            return arg.GetBoolean();
        }
        throw new ArgumentException("cannot convert " + arg.ValueKind + " to " + dest.Name);
    }

    static bool JsonEqual(object? actual, JsonElement expected)
    {
        return JsonSerializer.Serialize(actual) == JsonSerializer.Serialize(expected);
    }

    static Exception Unwrap(Exception exc)
    {
        if (exc is TargetInvocationException { InnerException: { } inner })
        {
            return inner;
        }
        return exc;
    }
}

File: langs/fsharp/Adapter.fs
module Adapter

open System
open System.Collections.Generic
open System.Reflection
open System.Text
open System.Text.Json

let private toCamel (snake: string) =
    let parts = snake.Split('_')
    let output = StringBuilder()
    let mutable first = true

    for part in parts do
        if part.Length > 0 then
            if first then
                output.Append(Char.ToLowerInvariant(part[0])) |> ignore
                first <- false
            else
                output.Append(Char.ToUpperInvariant(part[0])) |> ignore

            if part.Length > 1 then
                output.Append(part.Substring(1)) |> ignore

    output.ToString()

let rec private unwrap (exc: Exception) =
    match exc with
    | :? TargetInvocationException as tie when not (isNull tie.InnerException) -> unwrap tie.InnerException
    | _ -> exc

let private failRow (caseId: string) (index: int) (methodName: string) (expected: JsonElement) (actual: obj) =
    let row = Dictionary<string, obj>()
    row["case"] <- caseId
    row["index"] <- index
    row["method"] <- methodName
    row["expected"] <- expected
    row["actual"] <- actual
    row

let private findMethod (typ: Type) (name: string) (argc: int) =
    typ.GetMethods()
    |> Array.tryFind (fun method -> method.Name = name && method.GetParameters().Length = argc)

let private convertArg (arg: JsonElement) (dest: Type) : obj =
    let inner = Nullable.GetUnderlyingType(dest)
    let target = if isNull inner then dest else inner

    if arg.ValueKind = JsonValueKind.Null then
        if dest.IsValueType && isNull (Nullable.GetUnderlyingType(dest)) then
            raise (ArgumentException("null for non-nullable " + dest.Name))

        null
    elif target = typeof<string> then
        if arg.ValueKind <> JsonValueKind.String then
            raise (ArgumentException("cannot convert " + string arg.ValueKind + " to string"))

        box (arg.GetString())
    elif target = typeof<int> then
        if arg.ValueKind <> JsonValueKind.Number then
            raise (ArgumentException("cannot convert " + string arg.ValueKind + " to int"))

        box (arg.GetInt32())
    elif target = typeof<int64> then
        if arg.ValueKind <> JsonValueKind.Number then
            raise (ArgumentException("cannot convert " + string arg.ValueKind + " to long"))

        box (arg.GetInt64())
    elif target = typeof<bool> then
        if arg.ValueKind <> JsonValueKind.True && arg.ValueKind <> JsonValueKind.False then
            raise (ArgumentException("cannot convert " + string arg.ValueKind + " to bool"))

        box (arg.GetBoolean())
    else
        raise (ArgumentException("cannot convert " + string arg.ValueKind + " to " + dest.Name))

let private invoke (obj: obj) (name: string) (argv: JsonElement list) =
    match findMethod (obj.GetType()) name argv.Length with
    | None -> raise (MissingMethodException(name))
    | Some method ->
        let types = method.GetParameters()
        let converted =
            argv
            |> List.mapi (fun i arg -> convertArg arg types[i].ParameterType)
            |> List.toArray

        method.Invoke(obj, converted)

let private jsonEqual (actual: obj) (expected: JsonElement) =
    JsonSerializer.Serialize(actual) = JsonSerializer.Serialize(expected)

[<EntryPoint>]
let main (args: string[]) =
    let positional =
        args
        |> Array.toList
        |> List.filter (fun item -> item.Length > 0 && item <> "--" && not (item.StartsWith('-')))

    if positional.Length < 2 then
        eprintfn "usage: honepadrun cases.json ClassName"
        2
    else
        let casesPath = positional[0]
        let className = positional[1]
        use doc = JsonDocument.Parse(IO.File.ReadAllText(casesPath))

        if doc.RootElement.ValueKind <> JsonValueKind.Array then
            raise (InvalidOperationException("cases.json must be a JSON list"))

        let typ =
            Assembly.GetExecutingAssembly().GetTypes()
            |> Array.tryFind (fun item -> item.Name = className)
            |> Option.defaultWith (fun () -> raise (InvalidOperationException("missing type " + className)))

        let failed = ResizeArray<Dictionary<string, obj>>()
        let mutable passed = 0

        for row in doc.RootElement.EnumerateArray() do
            let instance = Activator.CreateInstance(typ)

            if isNull instance then
                raise (InvalidOperationException("could not construct " + className))

            let caseId =
                match row.GetProperty("id").GetString() with
                | null -> ""
                | value -> value

            let calls = row.GetProperty("calls")
            let mutable ok = true
            let mutable index = 0

            for call in calls.EnumerateArray() do
                if ok then
                    let methodSnake =
                        match call.GetProperty("m").GetString() with
                        | null -> ""
                        | value -> value

                    let name = toCamel methodSnake
                    let argv = call.GetProperty("a").EnumerateArray() |> Seq.toList
                    let expected = call.GetProperty("e")

                    try
                        let actual = invoke instance name argv

                        if not (jsonEqual actual expected) then
                            failed.Add(failRow caseId index methodSnake expected actual)
                            ok <- false
                    with exc ->
                        failed.Add(failRow caseId index methodSnake expected ("exc:" + (unwrap exc).GetType().Name))
                        ok <- false

                    if ok then
                        index <- index + 1

            if ok then
                passed <- passed + 1

        let report = Dictionary<string, obj>()
        report["passed"] <- passed
        report["failed"] <- failed
        printfn "%s" (JsonSerializer.Serialize(report))
        if failed.Count = 0 then 0 else 1
